harness_tests/diagnostic_test/main1.c

```diff
@@ -1,205 +1,454 @@
 /**
- * Diagnostic test to verify hardware functionality for EPS. This consists of: 
- *      - taking measurements with devices and making sure they are within 
- *          expected range
- *      - turning on a heater shuld increase current from boost and battery
- *      - imu gyro vals should be small numbers
- */
+ * Diagnostic test to verify hardware functionality for EPS.
+ *      - See Bus Architecture Specification (section 5- diagnostic tests) for more information
+ **/
+
 #include <stdbool.h>
 
 #include <adc/adc.h>
 #include <conversions/conversions.h>
 #include <uart/uart.h>
 #include <test/test.h>
+#include <can/data_protocol.h>
+
+#include "../../src/general.h"
 
 #include "../../src/devices.h"
 #include "../../src/heaters.h"
-#include "../../src/imu.h"
 #include "../../src/measurements.h"
-#include "../../src/shunts.h"
+#include "../../src/imu.h"
 
 
-double read_voltage(uint8_t channel) {
-    fetch_adc_channel(&adc, channel);
-    uint16_t raw_data = read_adc_channel(&adc, channel);
-    double voltage = adc_raw_to_circ_vol(raw_data, ADC_VOL_SENSE_LOW_RES, ADC_VOL_SENSE_HIGH_RES);
-    return voltage;
+uint8_t tx_msg[8] = {0x00};
+
+/* Helper function for generating and appropriately dequeueing
+    rx_housekeeping or rx_ctrl messages */
+uint32_t construct_rx_msg(uint8_t op_code, uint8_t field_num, uint32_t tx_data){
+    /* Initialize queue sizes to invalid valid */
+    uint8_t rx_q_size = -1;
+    uint8_t tx_q_size = -1;
+    uint8_t rx_msg[8] = {0x00};
+
+    rx_msg[2] = op_code;
+    rx_msg[3] = field_num;
+    tx_msg[4] = (tx_data >> 24) & 0xFF;
+    tx_msg[5] = (tx_data >> 16) & 0xFF;
+    tx_msg[6] = (tx_data >> 8) & 0xFF;
+    tx_msg[7] = tx_data & 0xFF;
+
+    enqueue(&can_rx_msg_queue, rx_msg);
+    rx_q_size = queue_size(&can_rx_msg_queue);
+    tx_q_size = queue_size(&can_tx_msg_queue);
+    ASSERT_EQ(rx_q_size, 1);
+    ASSERT_EQ(tx_q_size, 0);
+
+    handle_rx_msg();
+    rx_q_size = queue_size(&can_rx_msg_queue);
+    tx_q_size = queue_size(&can_tx_msg_queue);
+    ASSERT_EQ(rx_q_size, 0);
+    ASSERT_EQ(tx_q_size, 1);
+
+    dequeue(&can_tx_msg_queue, tx_msg);
+    rx_q_size = queue_size(&can_rx_msg_queue);
+    tx_q_size = queue_size(&can_tx_msg_queue);
+    ASSERT_EQ(rx_q_size, 0);
+    ASSERT_EQ(tx_q_size, 0);
```

Comment:
You should be getting the 4 bytes of data from `tx_msg` which you just dequeued and return that instead. The `tx_msg` array shouldn't need to be global.